bittide/tests/Tests/DoubleBufferedRam.hs

```diff
@@ -885,3 +891,169 @@ wbStorageProtocolsModel = property $ do
+ where
+  go :: forall wordsA wordsB m . (KnownNat wordsA, 2 <= wordsA, KnownNat wordsB, 2 <= wordsB, Monad m) => SNat wordsA -> SNat wordsB -> PropertyT m ()
+  go SNat SNat = do
+    wbRequestsNum <- forAll $ Gen.integral (Range.linearFrom 0 (0) 32)
+    wbRequestsA <- forAll $ Gen.list (Range.singleton wbRequestsNum)
+      (genWishboneTransfer @32 (natToNum @wordsA) (genDefinedBitVector @32))
+    wbRequestsB <- forAll $ Gen.list (Range.singleton wbRequestsNum)
+      (genWishboneTransfer @32 (natToNum @wordsB) (genDefinedBitVector @32))
+
+    cover 5 "equal address" $ or (L.zipWith checkAddressEqual (L.map fst wbRequestsA) (L.map fst wbRequestsB))
+
+    F.for_ (L.map (getAddressWbRequest . fst) (wbRequestsA L.++ wbRequestsB)) $ \x -> do
+        classify "low address" $ x == minBound
+        classify "mid address" $ x > minBound && x < maxBound
+        classify "high address" $ x == maxBound
+
+    F.for_ (L.map (getByteEnableWbRequest . fst) (wbRequestsA L.++ wbRequestsB)) $ \x -> do
+        classify "ByteEnable: 0000" $ x == 0
+        classify "ByteEnable: 0001" $ x == 1
+        classify "ByteEnable: 0010" $ x == 2
+        classify "ByteEnable: 0100" $ x == 4
+        classify "ByteEnable: 1000" $ x == 8
+
+    let
+      masterA = driveStandard defExpectOptions wbRequestsA
+      masterB = driveStandard defExpectOptions wbRequestsB
+      master = prod2C masterA masterB
+      slave = wbStorageTDPC @System @System @32 clockGen resetGen enableGen clockGen resetGen enableGen
+      (simTransactionsA, simTransactionsB) = exposeDoubleWbTransactions (Just 1000) master slave
+      (goldenTransactionsA, goldenTransactionsB) = L.unzip (wbStorageTDPBehaviorModel (fmap fst wbRequestsA) (fmap fst wbRequestsB))
+
+    footnote $ "simTransactionsA" <> show simTransactionsA
+    footnote $ "simTransactionsB" <> show simTransactionsB
+    footnote $ "wbRequestsA" <> show wbRequestsA
+    footnote $ "wbRequestsB" <> show wbRequestsB
+    footnote $ "goldenTransactionsA" <> show goldenTransactionsA
+    footnote $ "goldenTransactionsB" <> show goldenTransactionsB
+
+    assert $ and (L.zipWith compareTransWithUndef simTransactionsA goldenTransactionsA)
+    assert $ and (L.zipWith compareTransWithUndef simTransactionsB goldenTransactionsB)
+
+   where
+    genWishboneTransfer ::
+      (KnownNat addressWidth, KnownNat (BitSize a)) =>
+      Int -> -- ^ size
+      Gen a ->
+      Gen (WishboneMasterRequest addressWidth a, Int)
+    genWishboneTransfer size genA =
+      let
+        validAddr = (4 *) . fromIntegral <$> Gen.enum 0 (size - 1)
+        -- Make wbOps that won't be repeated
+        mkRead address bs = (Read address bs, 0)
+        mkWrite address bs a = (Write address bs a, 0)
+      in
+        -- Generate valid operations. The boolean represents the validity of the operation.
+      Gen.choice
+        [ (mkRead  <$> validAddr <*> genDefinedBitVector)
+        , (mkWrite <$> validAddr <*> genDefinedBitVector <*> genA)
+        ]
```

Comment:
I see I've been copying this around too.. Shame on me!

Can we make a generic, re-usable function for this?

Reply:
I noticed that this function occurs multiple times, with a slightly changed form. I will make a generic during refactoring.